File: csharp/OoBootCamp/Quantities/Unit.cs
/* 
 * Copyright (c) 2019 by Fred George
 * May be used freely except for training; license required for training.
 */

 using static OoBootCamp.Quantities.Unit;

namespace OoBootCamp.Quantities
{
    // Understands a specific metric
    public class Unit
    {
        public static readonly Unit Teaspoon = new Unit();
        public static readonly Unit Tablespoon = new Unit(3, Teaspoon);
        public static readonly Unit Ounce = new Unit(2, Tablespoon);
        public static readonly Unit Cup = new Unit(8, Ounce);
        public static readonly Unit Pint = new Unit(2, Cup);
        public static readonly Unit Quart = new Unit(2, Pint);
        public static readonly Unit Gallon = new Unit(4, Quart);

        private readonly double _baseUnitRatio;
        private Unit()
        {
            _baseUnitRatio = 1;
        }

        private Unit(double relativeRatio, Unit relativeUnit)
        {
            _baseUnitRatio = relativeRatio * relativeUnit._baseUnitRatio;
        }

        public Quantity S(double amount) => new Quantity(amount, this);

        internal double ConvertedAmount(double otherAmount, Unit other) => otherAmount * other._baseUnitRatio / this._baseUnitRatio;

        internal int HashCode(double amount) => (amount * _baseUnitRatio).GetHashCode();
    }

    namespace ExtensionMethods
    {
        public static class QuantityConstructors
        {
            public static Quantity Teaspoons(this double amount) => Teaspoon.S(amount);
            public static Quantity Teaspoons(this int amount) => Teaspoon.S(amount);
            public static Quantity Tablespoons(this double amount) => Tablespoon.S(amount);
            public static Quantity Tablespoons(this int amount) => Tablespoon.S(amount);
            public static Quantity Ounces(this double amount) => Ounce.S(amount);
            public static Quantity Ounces(this int amount) => Ounce.S(amount);
            public static Quantity Cups(this double amount) => Cup.S(amount);
            public static Quantity Cups(this int amount) => Cup.S(amount);
            public static Quantity Pints(this double amount) => Pint.S(amount);
            public static Quantity Pints(this int amount) => Pint.S(amount);
            public static Quantity Quarts(this double amount) => Quart.S(amount);
            public static Quantity Quarts(this int amount) => Quart.S(amount);
            public static Quantity Gallons(this double amount) => Gallon.S(amount);
            public static Quantity Gallons(this int amount) => Gallon.S(amount);

        }
    }

}

File: kotlin/exercises/src/quantity/Unit.kt
/*
 * Copyright (c) 2019 by Fred George
 * May be used freely except for training; license required for training.
 * @author Fred George  fredgeorge@acm.org
 */

package quantity

// Understands a specific metric
class Unit {
    companion object {
        internal val teaspoon = Unit()
        internal val tablespoon = Unit(3, teaspoon)
        internal val ounce = Unit(2, tablespoon)
        internal val cup = Unit(8, ounce)
        internal val pint = Unit(2, cup)
        internal val quart = Unit(2, pint)
        internal val gallon = Unit(4, quart)

        internal val inch = Unit()
        internal val foot = Unit(12, inch)
        internal val yard = Unit(3, foot)
        internal val chain = Unit(22, yard)
        internal val furlong = Unit(10, chain)
        internal val mile = Unit(8, furlong)
    }

    private val baseUnit: Unit
    private val baseUnitRatio: Double

    private constructor() {
        baseUnit = this
        baseUnitRatio = 1.0
    }

    private constructor(relativeRatio: Number, relativeUnit: Unit) {
        baseUnit = relativeUnit.baseUnit
        baseUnitRatio = relativeRatio.toDouble() * relativeUnit.baseUnitRatio
    }

    internal fun isCompatible(other: Unit) = this.baseUnit == other.baseUnit

    internal fun convertedAmount(otherAmount: Double, other: Unit): Double {
        require (this.isCompatible(other)) { "Incompatible Unit types" }
        return otherAmount * other.baseUnitRatio / this.baseUnitRatio
    }

    internal fun hashCode(amount: Double) = (amount * baseUnitRatio).hashCode()

}

val Number.teaspoons get() = Quantity(this, Unit.teaspoon)
val Number.tablespoons get() = Quantity(this, Unit.tablespoon)
val Number.ounces get() = Quantity(this, Unit.ounce)
val Number.cups get() = Quantity(this, Unit.cup)
val Number.pints get() = Quantity(this, Unit.pint)
val Number.quarts get() = Quantity(this, Unit.quart)
val Number.gallons get() = Quantity(this, Unit.gallon)

val Number.inch get(): Quantity = Quantity(this, Unit.inch)
val Number.inches get() = this.inch
val Number.foot get(): Quantity = Quantity(this, Unit.foot)
val Number.feet get() = this.foot
val Number.yards get(): Quantity = Quantity(this, Unit.yard)
val Number.chains get(): Quantity = Quantity(this, Unit.chain)
val Number.furlongs get(): Quantity = Quantity(this, Unit.furlong)
val Number.miles get(): Quantity = Quantity(this, Unit.mile)